src/traces/carpet/plot.js

```diff
@@ -0,0 +1,229 @@
+        var orientation;
+        if(label.axis.tickangle === 'auto') {
+            orientation = orientText(trace, xaxis, yaxis, label.xy, label.dxy);
+        } else {
+            var angle = (label.axis.tickangle + 180.0) * Math.PI / 180.0;
+            orientation = orientText(trace, xaxis, yaxis, label.xy, [Math.cos(angle), Math.sin(angle)]);
+        }
+        var direction = (label.endAnchor ? -1 : 1) * orientation.flip;
+        var bbox = measureText(tester, label.text, label.font);
+
+        d3.select(this)
+            .attr('text-anchor', direction > 0 ? 'start' : 'end')
+            .text(label.text)
+            .attr('transform',
+                // Translate to the correct point:
+                'translate(' + orientation.p[0] + ',' + orientation.p[1] + ') ' +
+                // Rotate to line up with grid line tangent:
+                'rotate(' + orientation.angle + ')' +
+                // Adjust the baseline and indentation:
+                'translate(' + label.axis.labelpadding * direction + ',' + bbox.height * 0.3 + ')'
+            )
+            .call(Drawing.font, label.font.family, label.font.size, label.font.color);
+
+        maxExtent = Math.max(maxExtent, bbox.width + label.axis.labelpadding);
+    });
+
+    labelJoin.exit().remove();
+
+    return maxExtent;
+}
+
+function drawAxisTitles(layer, trace, t, xa, ya, maxAExtent, maxBExtent) {
+    var a, b, xy, dxy;
+
+    a = 0.5 * (trace.a[0] + trace.a[trace.a.length - 1]);
+    b = trace.b[0];
+    xy = trace.ab2xy(a, b, true);
+    dxy = trace.dxyda_rough(a, b);
+    drawAxisTitle(layer, trace, t, xy, dxy, trace.aaxis, xa, ya, maxAExtent, 'a-title');
+
+    a = trace.a[0];
+    b = 0.5 * (trace.b[0] + trace.b[trace.b.length - 1]);
+    xy = trace.ab2xy(a, b, true);
+    dxy = trace.dxydb_rough(a, b);
+    drawAxisTitle(layer, trace, t, xy, dxy, trace.baxis, xa, ya, maxBExtent, 'b-title');
+}
+
+function drawAxisTitle(layer, trace, t, xy, dxy, axis, xa, ya, offset, labelClass) {
+    var data = [];
+    if(axis.title) data.push(axis.title);
+    var titleJoin = layer.selectAll('text.' + labelClass).data(data);
+
+    titleJoin.enter().append('text')
+        .classed(labelClass, true);
+
+    // There's only one, but we'll do it as a join so it's updated nicely:
+    titleJoin.each(function() {
+        var orientation = orientText(trace, xa, ya, xy, dxy);
+
+        if(['start', 'both'].indexOf(axis.showticklabels) === -1) {
+            offset = 0;
+        }
+
+        // In addition to the size of the labels, add on some extra padding:
+        offset += axis.titlefont.size + axis.titleoffset;
+
+
+        var el = d3.select(this);
+
+        el.text(axis.title || '')
+            .attr('transform',
+                'translate(' + orientation.p[0] + ',' + orientation.p[1] + ') ' +
+                'rotate(' + orientation.angle + ') ' +
+                'translate(0,' + offset + ')'
+            )
+            .classed('user-select-none', true)
```

Comment:
Ah, more things should perhaps have that. The double-click behavior (e.g. legend isolate) is kind of annoying when it selects a bunch of axis labels in the process.